crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -36,26 +38,73 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
         query: String,
         operation: Option<String>,
         options: QueryPlanOptions,
-    ) -> Result<Arc<QueryPlan>, QueryPlannerError> {
-        if let Some(value) = self
-            .cached
-            .lock()
-            .await
-            .get(&(query.clone(), operation.clone(), options.clone()))
-            .cloned()
-        {
+    ) -> PlanResult {
+        let mut locked_cache = self.cached.lock().await;
+        let key = (query, operation, options);
+        if let Some(value) = locked_cache.get(&key).cloned() {
             return value;
         }
 
-        let value = self
-            .delegate
-            .get(query.clone(), operation.clone(), options.clone())
-            .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
-        value
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        drop(locked_cache);
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                let msg = receiver.recv().await?;
```

Comment:
According to the [documentation](https://docs.rs/tokio/1.14.0/tokio/sync/broadcast/error/enum.RecvError.html) there are only 2 possible errors:

1. The channel was closed: "There are no more active senders implying no further messages will ever be sent."
2. "The receiver lagged too far behind. Attempting to receive again will return the oldest message still retained by the channel."

I guess point 2. makes no sense because only one message will ever be sent to this channel.

Point 1 happens only if all the senders are dropped but there is a receiver awaiting somewhere. This won't happen because the locked_wait_map will be locked while the message is being sent to the channel so no other receiver can be created during that time:

```rust
                // ...
                let mut locked_wait_map = self.wait_map.lock().await;
                locked_wait_map.remove(&key);
                // Let our waiters know
                let broadcast_value = value.clone();
                tokio::task::spawn_blocking(move || tx.send((key, broadcast_value))).await??;
                value
            } // locked_wait_map is dropped here
```

So I think you can safely do `.expect()` on this PR and explain the proof.

We don't want to add the RecvError to the QueryPlannerError as this can be returned to the user making the call. So either we make a strategy to "retry" to prevent failing or we unwrap because we are 100% confident that this will not fail. Correct me if I'm wrong but I think it's pretty clear that we can prove this won't ever fail.

Reply:
I agree that the documentation suggest this is safe to do, so I've done it.